Split-up Codebase into Smaller Chunks

mgerstner opened the pull request:
This addresses issue #59. It is mostly a split-up of the large Chkstat class into smaller, more managable pieces. Along with this comes some refactoring where it offered itself due to the new design.

mgerstner committed:
chkstat: drop SaneValueArg wrapper

Newer tclap versions available in TW now support const accessors so we
no longer need this wrapper class.

mgerstner committed:
chkstat: split off command line arguments from Chkstat main class

This relates to issue #59. Chkstat should be split into more manageable
pieces. Command line arguments are a good starting point.

The Chkstat class only receives a const reference to the new CmdlineArgs
type. The special SwitchArgRW wrapper is no longer needed since we keep
a regular bool in Chkstat for keeping track of the runtime
`apply_changes` status. The purely logical checking of command line
arguments is kept in the new CmdlineArgs type while actual processing
stays in Chkstat itself.

mgerstner committed:
chkstat: replace #include guards by #pragma once

The pragma is less error prone, less noisy and also more efficient for
compilation times. Practically all compilers support it.

mgerstner committed:
chkstat: move assorted types into dedicated header

mgerstner committed:
chkstat: remove deprecated CHECK_PERMISSIONS logic

mgerstner committed:
chkstat: parseSysconfig(): bit of refactoring

mgerstner committed:
chkstat: loadVariableExpansions(): a bit of refactoring

mgerstner committed:
chkstat: split off variable expansion logic into separate class

mgerstner committed:
chkstat: also move expandProfilePaths() into new VariableExpansions class

mgerstner committed:
chkstat: drop deprecated capability check

mgerstner committed:
chkstat: parseProfile(): improve badProfileLine() calls

mgerstner committed:
chkstat: ProfileEntry: mark dropXID() const to support const ProfileEntry use

mgerstner committed:
chkstat: split off separate ProfileParser

This moves the actual permissions profile parser logic into its own
class and thereby further simplifies the main Chkstat class.

mgerstner committed:
chkstat: ProfileParser: refactor the now reduced codebase

The complexity can be managed better by adding more separate functions
and by keeping the current parsing context as class state.

mgerstner committed:
chkstat: ProfileParser: make adding the root to paths transparent

Now that we have a separate ProfileParser we can also hide the
implementation detail of how paths are kept in the entries map.

mgerstner committed:
chkstat: ProfileParser: fix a bug when applying capabilities in custom root

The code never added a potential `-r` root path to the std::map key i.e.
if capabilities are used together with an alternative root directory
then parsing will never succeed.

Adjust the test suite to catch this in the future. This test fails
without the applied patch and works after the patch.

mgerstner committed:
chkstat: split up checkHaveProc()

The dual nature of checkHaveProc() is problematic which can be seen from
the mutable state of the m_proc_mount_avail. Split this function into a
dedicated procState() method that performs the system calls and a pure
getter function that can be const.

mgerstner committed:
chkstat: processEntries(): make loop variables const

With ProfileEntry::dropXID() being const this can now also be const.

mgerstner committed:
chkstat: define _GNU_SOURCE via Makefile

Doing this in the source bloats up the code and risks inconsistencies.

mgerstner committed:
chkstat: split-off EntryProcessor from Chkstat main class

This new type holds the actual file system logic needed for applying a
specific permissions entry. This split reduces the main Chkstat class to
a holder of configuration items, some global policy decisions and
forwarder of information between sub entities.

This also simplifies the member function signatures of EntryProcessor in
some places since the class now holds an EntryContext and ProfileEntry
as class members.

The current Chkstat::processEntries() is very selective in what
situation is counted as an error to report to the caller and what not.
While splitting up this now counts more situations as errors. Let's hope
this doesn't blow up.

mgerstner committed:
chkstat: get rid of EntryContext and incorporate it into EntryProcessor

The EntryContext is now mostly noise, so incorporate it as class members
of EntryProcessor.

mgerstner committed:
chkstat: EntryProcessor: rename some member variables for improved readability

mgerstner committed:
chkstat: make ProcMountState a private type of ChkStat

mgerstner committed:
chkstat: drop types.h header

With the now clearer split-up responsibilities there's no need for an
assorted types header anymore. Move the remaining ProfileEntry into
profparser.h, where it belongs semantically, and drop the types.h
header.

mgerstner committed:
chkstat: minor style, spelling and documentation fixes

FilippoBonazziSUSE commented:
I will continue reviewing from [chkstat: ProfileParser: refactor the now reduced codebase](https://github.com/openSUSE/permissions/pull/184/commits/c610e094d650b4fb4d03115ceda0c20e619b9e97) tomorrow

FilippoBonazziSUSE commented on src/chkstat.cpp:
```diff
@@ -18,25 +18,19 @@
  ****************************************************************
  */
 
-#include <dirent.h>
```
`<dirent.h> is just dropped. is it no longer necessary?

mgerstner replied:
I'm not sure why this was still there. Maybe `opendir()` was used in the past. But now it's no longer necessary.

FilippoBonazziSUSE commented on src/entryproc.h:
```diff
@@ -0,0 +1,74 @@
+    static const gid_t m_egid; ///< The effective group ID we're running as.
+};
+
+// vim: et ts=4 sts=4 sw=4 :
```
Very large commit, but should be ok. I did not look further into the new error conditions for `processEntries` compared to the pre-refactor implementation, if there is cause for concern perhaps this is an area where tests could help.

FilippoBonazziSUSE commented on src/entryproc.h:
```diff
@@ -58,12 +58,31 @@ class EntryProcessor {
+    bool m_need_fix_ownership = false; ///< Indicates whether actual file user:group ownership needs to be fixed.
+    bool m_traversed_insecure = false; ///< An insecure ownership constellation was detected in the file's path.
+    FileDesc m_fd; ///< An `O_PATH` file descriptor for the target file which is set in safeOpen().
+
```
Ok

FilippoBonazziSUSE commented on src/entryproc.h:
```diff
@@ -74,9 +74,9 @@ class EntryProcessor {
     gid_t m_file_gid = (gid_t)-1; ///< The resolved group-id corresponding to the active ProfileEntry.
     std::string m_path; ///< The path of the current file to check below a potential m_args.root_path.
-    std::string m_fd_path; ///< A path for safely opening the target file (typically in /proc/self/fd/...).
+    std::string m_safe_path; ///< A path for safely opening the target file (typically in /proc/self/fd/...).
```
Why "typically"? Are there circumstances where this is not true?

mgerstner replied:
Yes in diff / read-only mode, if /proc is not available, as is documented elsewhere in the code.

FilippoBonazziSUSE replied:
In that case, is `m_safe_path` not set or is it set to something else?

mgerstner replied:
There are three levels and one sub varaint of paths here:

- `m_entry.file` is the full path, which could be including an "alternative root".
- `m_path` is the display path to show for diagnostics, a human readable version, relative to a possible "alternative root".
- `m_safe_path` is an implementation detail that usually refers to a safe variant of the path pointing to /proc.
-  sub variant of `m_safe_path`: an "unsafe" path if `chkstat` considers it okay.

It comes with the heritage of this program that generic code portions are reused for vastly different contexts. I guess it is hard to change this at this stage. We have the feature to operate only in a checking read-only mode where using safe paths is not strictly required. And we have this friendly fallback to allow this read-only mode even if there is no `/proc`.

Basically the codebase guarantees via the logic in `EntryProcess::process()` that `m_safe_path` always is safe given the context that was detected. So it could be argued that it _is_ always safe.

To make this clearer see different possibilities:

- drop the fallback to operate on "unsafe" paths if there is no `/proc` and we only want to "check". I don't believe this feature will ever come into play. `/proc` not being available usually only happens in build environments, as we have seen lately, and I guess no one will enter the build environment to "check" which permissions are missing while `/proc` is not mounted.
- choose a clearer name. Maybe something like `m_api_path` to emphasize the property that it is a variant of the path specific to the APIs used.
- keep it as it is, refering to the reasoning above.

FilippoBonazziSUSE commented on src/chkstat.h:
```diff
@@ -22,6 +22,16 @@ class Chkstat {
 
     int run();
 
+protected: // types
```
Ok

FilippoBonazziSUSE commented on src/profparser.h:
```diff
@@ -2,14 +2,41 @@
 #include <map>
 #include <string>
 
+/// Represents a single permissions profile entry.
```
Ok

FilippoBonazziSUSE reviewed (changes requested):
Massive changes. Ok overall, many comments to address or to discuss.

Maybe we can schedule a separate session where we go over the whole program as it is after the refactoring, and you explain to me the logic and program flow. I feel like I would benefit from that and we can also probably still find something to refactor/harmonize. Let's discuss this separately.

mgerstner commented on src/chkstat.cpp:
```diff
@@ -646,14 +571,14 @@ void Chkstat::printHeader() {
 
 void Chkstat::printEntryDifferences(const ProfileEntry &entry, const EntryContext &ctx) const {
     std::cout << ctx.subpath << ": "
-        << (m_apply_changes.isSet() ? "setting to " : "should be ")
```
For a `TCLAP::SwitchArg` which is either present or not without an actual value on the command line there is no difference between `getValue()` and `isSet()`.

The `should be` is used when changes are not to be applied but when `chkstat` should only display what it _would_ do.

mgerstner commented on src/chkstat.cpp:
```diff
@@ -1092,15 +1017,12 @@ int Chkstat::processEntries() {
     loadVariableExpansions();
 
-    if (m_print_variables.isSet()) {
+    if (m_args.print_variables.isSet()) {
```
It's the same as above, for `TCLAP::SwitchArg` `isSet()` and `getValue()` are equivalent.

mgerstner commented on src/cmdline.cpp:
```diff
@@ -0,0 +1,79 @@
+int CmdlineArgs::parse(const int argc, const char **argv) {
+    m_parser.parse(argc, argv);
+
+    return validateArguments() ? 0 : 2;
```
Process exit codes are rather arbitrary unless they are specifically documented. Except that `0` is generally interpreted as successful termination of the process.

What is attempted here, I guess, is differentiating between a usage error (i.e. wrong command line arguments supplied) and an operational error of any other kind. To make this clearer constant can be introduced. If you want that then I would add this as another commit on top of the PR. Ideally this would also be documented in the man page than. But I doubt that anybody would be evaluating that. It could rather be a debugging aid if `chkstat` fails in whatever circumstances and we can only see the exit code in the logs, for example.

mgerstner commented on src/entryproc.h:
```diff
@@ -23,6 +23,11 @@ class EntryProcessor {
+    /// Perform the necessary processing logic.
+    /**
+     * \param[in] have_proc Whether a /proc mount is available. If not then
+     * not changes will be applied in the file system for security reasons.
```
I can rebase or first put a fixing commit on top to make it easier to review.

mgerstner commented on src/chkstat.cpp:
```diff
@@ -1033,6 +1007,8 @@ int Chkstat::run() {
 
     if (m_args.system_mode.isSet()) {
+        // this overrides --set
+        m_apply_changes = m_args.only_warn.isSet() ? false : true;
```
Yes this is what it boils down to. Previously this read-only mode could also be enabled via sysconfig, or rather, the applying of changes could be disabled. Which is confusing on many levels, because `if CHECK_PERMISSIONS == "true"` then permissions would not only be checked but also be set, while `if CHECK_PERMISSIONS == "false"` then `chkstat` would not continue running at all. Time for a headache pill.

mgerstner commented on src/types.h:
```diff
@@ -29,7 +29,8 @@ struct ProfileEntry {
     }
 
-    void dropXID() {
+    // NOTE: this is currently mutable due to the somewhat unfortunate logic in Chkstat::getCapabilities().
```
Ah yes, this `mutable` refers to the mutable variable actually that makes this `const` method possible. I can shift the comment.

mgerstner replied:
I pushed a commit for this on top that should be rebased once we are finished with the review.

mgerstner commented on src/chkstat.cpp:
```diff
@@ -516,7 +370,7 @@ void Chkstat::printEntryDifferences(const ProfileEntry &entry, const EntryContex
 }
 
-bool Chkstat::getCapabilities(ProfileEntry &entry, EntryContext &ctx) {
+bool Chkstat::getCapabilities(const ProfileEntry &entry, EntryContext &ctx) {
```
Due to the `dropXID()` being const now (although is modified the mutable member variable), the structure can now be declared as `const`. This way the compiler can at least guarantee that none of the unmutable member variables is accidentally modified.

mgerstner commented on src/profparser.h:
```diff
@@ -65,6 +68,20 @@ class ProfileParser {
+            mode = 0;
+        }
+    } m_parse_context;
+
```
I've run the regression tests for each commit so that gives us some confidence.

mgerstner commented on src/utility.h:
```diff
@@ -49,6 +49,12 @@ void strip(std::string &s, UNARY f = chkspace) {
 }
 
+template <typename UNARY = bool(char)>
+std::string stripped(std::string s, UNARY f = chkspace) {
```
Since I'm calling `strip()` here I need to use the same template declaration it has. I guess this could also be done without a template these days. By using the template this could also be used with a different function signature that somehow evalutes to bool. Although I don't believe this feature is used in the codebase currently.

mgerstner commented on src/chkstat.cpp:
```diff
@@ -298,19 +298,21 @@ void Chkstat::collectPackageProfilePaths(const std::string &dir) {
+    struct statfs proc;
+    int r = statfs("/proc", &proc);
+
+    if (r == 0 && proc.f_type == PROC_SUPER_MAGIC) {
```
I've seen worse :-P

It simply means the system call succeeded i.e. `proc` has been filled in by the kernel and now we check if the file system type actually is for the proc-fs.

FilippoBonazziSUSE replied:
I was thrown off by the super magic bit. But that's just how statfs is. Shows my inexperience, my bad

mgerstner commented on src/chkstat.cpp:
```diff
@@ -682,7 +684,9 @@ std::string Chkstat::getPathFromProc(const FileDesc &fd) const {
     size_t errors = 0;
 
-    if (m_apply_changes && !checkHaveProc()) {
+    m_proc_mount_avail = procState();
```
There's no TOCTOU at least no new one compared to the original code. Of course /proc could disappear or appear in the meanwhile but that is not a scenario that the tool is concerned with. I don't believe it would be possible to protect against a parallel mount/unmount operation unless we split off a dedicated mount namespace and mount proc on our own.

I already thought about getting rid of the `ProcMountState` enum altogether and just keep a boolean around that gets initialized early on. This could be done in the Chkstat constructor already, thereby keeping `m_proc_mount_avail` const. The only downside to this is that the check would happen even if it is unnecessary e.g. when running something like `chkstat --print-variables`.

FilippoBonazziSUSE replied:
Ok, perhaps worrying about TOCTOU is a bit too sci-fi. Two things could happen:
1. `proc` is not mounted when we check, and is surreptitiously mounted later:
    In this case the tool will just operate in a reduced capacity, plainly visible to the user, and will not do any damage.
2. `proc` is mounted when we check, and is surreptitiously unmounted later:
    In this case the tool will most likely just crash upon first access of a path under `proc`, also plainly visible to the user. Probably no damage is done in this case either :thinking: 

As for the `ProcMountState`, leaving only a boolean would indeed be simpler. And I don't think checking this unconditionally would add a significant overhead. Your call whether it's worth it or if leaving things as they are suffices.

mgerstner committed:
chkstat: replace ProcMountState enum by simple bool member

This now checks during construction time already whether /proc is
mounted or not. This allows to make the status flag const and we never
have an UNKNOWN state in the first place. A small downside to this is
that the logic now also runs even if it isn't technically required (e.g.
when --print-variables is passed). This is a sensible trade-off though.

FilippoBonazziSUSE reviewed (approved):
All comments addressed. LGTM

FilippoBonazziSUSE reviewed (approved):
Only rebase since last review - diff identical
